File: app/src/main/java/com/marknkamau/justjava/activities/profile/PreviousOrderAdapter.java
package com.marknkamau.justjava.activities.profile;

import android.content.Context;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.marknkamau.justjava.R;
import com.marknkamau.justjava.models.PreviousOrder;

import java.text.DateFormat;
import java.util.Calendar;
import java.util.Date;
import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;

public class PreviousOrderAdapter extends RecyclerView.Adapter<PreviousOrderAdapter.ViewHolder> {

    @BindView(R.id.tv_timestamp)
    TextView tvTimestamp;
    @BindView(R.id.tv_total_price)
    TextView tvTotalPrice;
    @BindView(R.id.tv_delivery_address)
    TextView tvDeliveryAddress;
    @BindView(R.id.tv_order_status)
    TextView tvOrderStatus;

    private Context mContext;
    private List<PreviousOrder> previousOrders;

    PreviousOrderAdapter(Context mContext, List<PreviousOrder> previousOrders) {
        this.mContext = mContext;
        this.previousOrders = previousOrders;
    }

    @Override
    public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        LayoutInflater inflater = LayoutInflater.from(mContext);
        View itemView = inflater.inflate(R.layout.item_previous_order, parent, false);
        ButterKnife.bind(this, itemView);

        return new ViewHolder(itemView);
    }

    @Override
    public void onBindViewHolder(ViewHolder holder, int position) {
        final PreviousOrder order = previousOrders.get(position);

        tvTimestamp.setText(getFormattedDate(Long.valueOf(order.getTimestamp())));
        tvTotalPrice.setText(String.format("%s%S", mContext.getString(R.string.ksh), order.getTotalPrice()));
        tvOrderStatus.setText(order.getOrderStatus());
        tvDeliveryAddress.setText(order.getDeliveryAddress());
    }

    private String getFormattedDate(long timestamp) {
        try {
            Calendar calendar = Calendar.getInstance();
            calendar.setTimeInMillis(timestamp);
            Date currentTime = calendar.getTime();
            return DateFormat.getDateTimeInstance().format(currentTime);
        } catch (Exception e) {
            e.printStackTrace();
        }
        return String.valueOf(timestamp);
    }

    @Override
    public int getItemCount() {
        return previousOrders.size();
    }

    class ViewHolder extends RecyclerView.ViewHolder {
        ViewHolder(View itemView) {
            super(itemView);
        }
    }
}

File: app/src/main/java/com/marknkamau/justjava/ui/profile/PreviousOrderAdapter.kt
package com.marknkamau.justjava.ui.profile

import android.annotation.SuppressLint
import android.content.Context
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.marknkamau.justjava.R
import com.marknkamau.justjava.data.models.Order
import com.marknkamau.justjava.utils.formatForApp
import kotlinx.android.synthetic.main.item_previous_order.view.*

class PreviousOrderAdapter(private val context: Context, private val onClick: (order: Order) -> Unit) : RecyclerView.Adapter<PreviousOrderAdapter.ViewHolder>() {
    private val items by lazy { mutableListOf<Order>() }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int) = ViewHolder(parent.inflate(R.layout.item_previous_order))

    override fun onBindViewHolder(holder: ViewHolder, position: Int) = holder.bind(items[position], context, onClick)

    override fun getItemCount() = items.size

    fun setItems(items: MutableList<Order>) {
        this.items.clear()
        this.items.addAll(items)
        notifyDataSetChanged()
    }

    class ViewHolder(itemView: View?) : RecyclerView.ViewHolder(itemView) {

        @SuppressLint("SetTextI18n")
        fun bind(order: Order, context: Context, onClick: (order: Order) -> Unit) {
            itemView.tvTimestamp.text = order.date.formatForApp()
            itemView.tvStatus.text = order.status.name.toLowerCase().capitalize()
            itemView.tvAddress.text = order.deliveryAddress
            itemView.tvOrderInfo.text = context.resources.getQuantityString(R.plurals.order_info, order.itemsCount, order.itemsCount, order.totalPrice)

            itemView.rootLayout.setOnClickListener {
                onClick(order)
            }
        }
    }


    private fun ViewGroup.inflate(layoutRes: Int): View {
        return LayoutInflater.from(context).inflate(layoutRes, this, false)
    }
}
